base/utf32.jl

```diff
@@ -20,15 +20,234 @@ length(s::UTF32String) = length(s.data) - 1
 function utf32(c::Integer...)
     a = Array(Char, length(c) + 1)
     for i = 1:length(c)
-        a[i] = Char(c[i])
+        @inbounds a[i] = Char(c[i])
     end
     a[end] = Char(0)
     UTF32String(a)
 end
 
+#=
+"""
+@brief      Converts a UTF-8 encoded string to UTF-32 encoding
+
+@param[in]  iStr::UTF8String
+
+@return     ::UTF32String
+@throws     UnicodeError
+
+            Hopefully copy! can be sped up to handle this case of widening
+"""
+=#
+function ascii_to_utf32(s::Array{UInt8,1})
+    len::Int = sizeof(s)
+    buf = Array{Char,1}(len+1)
+    for i = 1:len
+        @inbounds buf[i] = Char(s[i])
+    end
+    @inbounds buf[len+1] = Char(0) # NULL termination
+    UTF32String(buf)
+end
+
+#=
+"""
+@brief      Converts a UTF-8 encoded string to UTF-32 encoding
+
+@param[in]  iStr::UTF8String
+
+@return     ::UTF32String
+@throws     UnicodeError
+"""
+=#
+function utf8_to_utf32(iStr::Array{UInt8,1})
+    len::UInt = sizeof(iStr)
+    # handle zero length string quickly
+    if len == 0
+        buf = Array{Char,1}(1)
+        @inbounds buf[1] = Char(0)
+        return UTF32String(buf)
+    end
+    # get number of words to create
+    len, cnt2::UInt, cnt3::UInt, cnt4::UInt, flags::UInt = utf8_validate(iStr)
+    buf = Array{Char,1}(len+1)
+    @inbounds begin
+    buf[len+1] = Char(0) # NULL termination
+    if flags == 0
+        # 7-bit ASCII characters only! Do optimized conversion
+        for pos=1:len
+            buf[pos] = Char(iStr[pos])
+        end
+    else
+        # has multi-byte UTF-8 sequences
+        out::UInt = 0
+        pos::UInt = 0
+        while out < len
+            ch::UInt32 = iStr[pos += 1]
+            if ch > 0x7f
+                if ch < 0xe0
+                    ch = ((ch & 0x1f) << 6) | (iStr[pos += 1] & 0x3f)
+                elseif ch < 0xf0
+                    ch = ((ch & 0xf) << 12) | (UInt32(iStr[pos+1] & 0x3f) << 6) | (iStr[pos+2] & 0x3f)
+                    pos += 2
+		    if is_surrogate_lead(ch)
+			ch = (((ch & 0x3ff) << 10)
+				| (((UInt32(iStr[pos+1] & 0xf) << 12)
+				|  (UInt32(iStr[pos+2] & 0x3f) << 6)
+				|  (iStr[pos+3] & 0x3f)) & 0x3ff)) + 0x10000
+			pos += 3
+		    end
+                else
+                    ch = ((ch & 0x7) << 18) | (UInt32(iStr[pos+1] & 0x3f) << 12) | (UInt32(iStr[pos+2] & 0x3f) << 6) | (iStr[pos+3] & 0x3f)
+                    pos += 3
+                end
+            end
+            buf[out += 1] = Char(ch)
+        end
+    end
+    end
+    UTF32String(buf)
+end
+
+#=
+"""
+@brief      Converts a UTF-16 encoded string to UTF-32 encoding
+
+@param[in]  iStr::UTF16String
+
+@return     ::UTF32String
+@throws     UnicodeError
+"""
+=#
+function utf16_to_utf32(iStr::Array{UInt16,1})
```

Comment:
yep, i'm agreeing with you. the gist (https://gist.github.com/ScottPJones/bb712f7b85d1d8d91a9a) is lacking a simply summary view to see at a glance which combination of methods and inputs are being affected by this PR and by how much, so i'm providing a limited subset here.